File: src/0148_sort_list.js
function ListNode(val = 0, next = null) {
    this.val = val;
    this.next = next;
}

function sortList(head) {
    if (!head || !head.next) return head;
    const mid = getMid(head);
    const left = sortList(head);
    const right = sortList(mid);
    return mergeTwoSorted(left, right);
}

function mergeTwoSorted(list1, list2) {
    if (!list1) return list2;
    if (!list2) return list1;
    const sentinel = new ListNode();
    let prev = sentinel;
    while (list1 && list2) {
        if (list1.val < list2.val) {
            prev.next = list1;
            list1 = list1.next;
        } else {
            prev.next = list2;
            list2 = list2.next;
        }
        prev = prev.next;
    }
    prev.next = list1 || list2;
    return sentinel.next;
}

function getMid(head) {
    let midPrev = null;
    while (head && head.next) {
        midPrev = midPrev ? midPrev.next : head;
        head = head.next.next;
    }
    const mid = midPrev.next;
    midPrev.next = null;
    return mid;
}

module.exports = { sortList };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(listNodeToString(sortList(new ListNode(4, new ListNode(2, new ListNode(1, new ListNode(3, null)))))), listNodeToString(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, null))))));
    assert.deepStrictEqual(listNodeToString(sortList(new ListNode((-1), new ListNode(5, new ListNode(3, new ListNode(4, new ListNode(0, null))))))), listNodeToString(new ListNode((-1), new ListNode(0, new ListNode(3, new ListNode(4, new ListNode(5, null)))))));
    assert.deepStrictEqual(sortList(null), null);
}

File: src/0148_sort_list.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

struct Solution;

impl Solution {
    pub fn sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut values = list_to_vec(&head);
        merge_sort(&mut values);
        vec_to_list(values)
    }
}

fn merge_sort(values: &mut [i32]) {
    let len = values.len();
    if len <= 1 {
        return;
    }
    let mid = len / 2;
    merge_sort(&mut values[..mid]);
    merge_sort(&mut values[mid..]);

    let mut merged = Vec::with_capacity(len);
    let (left, right) = values.split_at(mid);
    let mut left_index = 0;
    let mut right_index = 0;
    while left_index < left.len() && right_index < right.len() {
        if left[left_index] < right[right_index] {
            merged.push(left[left_index]);
            left_index += 1;
        } else {
            merged.push(right[right_index]);
            right_index += 1;
        }
    }
    merged.extend_from_slice(&left[left_index..]);
    merged.extend_from_slice(&right[right_index..]);
    values.copy_from_slice(&merged);
}

fn list_to_vec(head: &Option<Box<ListNode>>) -> Vec<i32> {
    let mut values = Vec::new();
    let mut current = head.as_ref();
    while let Some(node) = current {
        values.push(node.val);
        current = node.next.as_ref();
    }
    values
}

fn vec_to_list(values: Vec<i32>) -> Option<Box<ListNode>> {
    let mut head = None;
    for value in values.into_iter().rev() {
        let mut node = Box::new(ListNode::new(value));
        node.next = head;
        head = Some(node);
    }
    head
}

#[cfg(test)]
mod tests {
    use super::*;

    fn list(values: &[i32]) -> Option<Box<ListNode>> {
        values
            .iter()
            .rev()
            .fold(None, |next, &val| Some(Box::new(ListNode { val, next })))
    }

    fn list_values(head: &Option<Box<ListNode>>) -> Vec<i32> {
        let mut values = Vec::new();
        let mut current = head.as_deref();
        while let Some(node) = current {
            values.push(node.val);
            current = node.next.as_deref();
        }
        values
    }

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            list_values(&Solution::sort_list(list(&[4, 2, 1, 3]))),
            vec![1, 2, 3, 4]
        );
        assert_eq!(
            list_values(&Solution::sort_list(list(&[-1, 5, 3, 4, 0]))),
            vec![-1, 0, 3, 4, 5]
        );
        assert!(Solution::sort_list(None).is_none());
    }
}
